internal/handlers/common/saslstart_plain.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/FerretDB/FerretDB/internal/handlers/commonerrors"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// saslStartPlain extracts username and password from PLAIN `saslStart` payload.
+func saslStartPlain(doc *types.Document) (string, string, error) {
+	var payload []byte
+
+	// some drivers send payload as a string
+	stringPayload, err := GetRequiredParam[string](doc, "payload")
+	if err == nil {
+		if payload, err = base64.StdEncoding.DecodeString(stringPayload); err != nil {
+			return "", "", lazyerrors.Error(err)
+		}
+	}
+
+	// most drivers follow spec and send payload as a binary
+	binaryPayload, err := GetRequiredParam[types.Binary](doc, "payload")
```

Comment:
That's a part I don't understand. A few lines above, there was a call `GetRequiredParam[string](doc, "payload")`. Let's say as a result of this call, `err` was nil, and even after `base64.StdEncoding.DecodeString` it was still `nil`. Then by this moment we will have some payload. Why do we need to call `GetRequiredParam[types.Binary](doc, "payload")` if we already have payload from `string`?

Reply:
That was introduced [here](https://github.com/FerretDB/FerretDB/pull/2181). Apparently the response can contain a base64 encoded string or a binary payload, see [spec](https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#sasl-mechanisms).

> payload: a sequence of bytes or a base64 encoded string (depending on input) to pass into the SASL library to transition the state machine.

Reply:
You say "or", but the code rather says "and" :)

Reply:
That's not correct – the code in this function does not expect this field to somehow have two types simultaneously

Reply:
Please give it 5 minutes

Reply:
@AlekSi I spent much more than 5 minutes and I still don't understand why we try to read binary payload in case we already successfully got a string payload and no errors were returned.

Reply:
Maybe we have an example of such a payload?

Reply:
I think you could try to rewrite that code to make it simpler with fewer branches

Reply:
Sure, how can I reproduce payloads?

Reply:
Please give 5 minutes :)